doc/apt.7
.\" This manpage is copyright (C) 1998 Branden Robinson <branden@debian.org>.
.\" 
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with APT; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 
.\" 02111-1307 USA
.TH apt 7 "16 June 1998" "Debian"
.SH NAME
apt \- Advanced Package Tool
.SH SYNOPSIS
.B apt
.SH DESCRIPTION
APT is a management system for software packages. 

For normal day to day
package management there are several frontends available, such as
.BR aptitude (8)
for the command line or
.BR synaptic (8)
for the X Window System. Some options are only implemented in
.BR apt-get (8)
though.
.SH SEE ALSO
.BR apt (8),
.BR apt-cache (8),
.BR apt-get (8),
.BR apt.conf (5),
.BR sources.list (5),
.BR apt_preferences (5),
.BR apt-secure (8)
.SH DIAGNOSTICS
apt returns zero on normal operation, decimal 100 on error.
.SH BUGS
This manpage isn't even started.
.PP
See <http://bugs.debian.org/apt>.  If you wish to report a
bug in
.BR apt ,
please see
.I /usr/share/doc/debian/bug-reporting.txt
or the
.BR reportbug (1)
command.
.SH AUTHOR
apt was written by the APT team <apt@packages.debian.org>.


debian/NEWS
apt (0.8.11) unstable; urgency=low

  * apt-get install pkg/experimental will now not only switch the
    candidate of package pkg to the version from the release experimental
    but also of all dependencies of pkg if the current candidate can't
    satisfy a versioned dependency.

 -- David Kalnischkies <kalnischkies@gmail.com>  Fri, 03 Dec 2010 14:09:12 +0100

apt (0.7.26~exp3) experimental; urgency=low

  * apt-ftparchive now reads the standard configuration files in
    /etc/apt/apt.conf and /etc/apt/apt.conf.d.

 -- Julian Andres Klode <jak@debian.org>  Fri, 26 Mar 2010 15:34:16 +0100

apt (0.7.24) unstable; urgency=low

  * Already included in the last version but now with better documentation
    is the possibility to add/prefer different compression types while
    downloading archive information, which can decrease the time needed for
    update on slow machines. See apt.conf (5) manpage for details.
  * APT manages his manpage translations now with po4a, thanks to Nicolas
    François and Kurasawa Nozomu, who also provide the ja translation.
    Thanks to Christian Perrier we have already a fr translation and
    a few more are hopefully added in the near future.
  * This version also introduces some _experimental_ configuration options
    to make more aggressive use of dpkg's triggers. If you want to help
    testing these _experimental_ options see apt.conf (5) manpage.

 -- David Kalnischkies <kalnischkies@gmail.com>  Thu, 24 Sep 2009 15:13:16 +0200

apt (0.7.23) unstable; urgency=low

  * Code that determines which proxy to use was changed. Now
    'Acquire::{http,ftp}::Proxy[::<host>]' options have the highest priority,
    and '{http,ftp}_proxy' environment variables are used only if options
    mentioned above are not specified.

 -- Eugene V. Lyubimkin <jackyf.devel@gmail.com>  Thu, 19 Aug 2009 11:26:16 +0200

apt (0.6.44) unstable; urgency=low

  * apt-ftparchive --db now uses Berkeley DB_BTREE instead of DB_HASH. 
    If you use a database created by an older version of apt, delete 
    it and allow it to be recreated the next time.

 -- Michael Vogt <mvo@debian.org>  Wed, 26 Apr 2006 12:57:53 +0200

apt (0.5.25) unstable; urgency=low

  * apt-ftparchive --db now uses Berkeley DB version 4.2.  If used with a
    database created by an older version of apt, an attempt will be made
    to upgrade the database, but this may not work in all cases.  If your
    database is not automatically upgraded, delete it and allow it to be
    recreated the next time.

 -- Matt Zimmerman <mdz@debian.org>  Sat,  8 May 2004 12:38:07 -0700



cmdline/apt.cc
// -*- mode: cpp; mode: fold -*-
// Description								/*{{{*/
/* ######################################################################
   
   apt - CLI UI for apt
   
   Returns 100 on failure, 0 on success.
   
   ##################################################################### */
									/*}}}*/
// Include Files							/*{{{*/
#include<config.h>

#include <apt-pkg/cmndline.h>
#include <apt-pkg/error.h>
#include <apt-pkg/init.h>
#include <apt-pkg/pkgsystem.h>
#include <apt-pkg/strutl.h>
#include <apt-pkg/configuration.h>

#include <apt-private/private-list.h>
#include <apt-private/private-search.h>
#include <apt-private/private-install.h>
#include <apt-private/private-output.h>
#include <apt-private/private-update.h>
#include <apt-private/private-cmndline.h>
#include <apt-private/private-moo.h>
#include <apt-private/private-upgrade.h>
#include <apt-private/private-show.h>
#include <apt-private/private-main.h>
#include <apt-private/private-sources.h>

#include <unistd.h>
#include <iostream>
#include <vector>

#include <apti18n.h>
									/*}}}*/

static bool ShowHelp(CommandLine &)
{
   ioprintf(c1out,_("%s %s for %s compiled on %s %s\n"),PACKAGE,PACKAGE_VERSION,
	    COMMON_ARCH,__DATE__,__TIME__);

   // FIXME: generate from CommandLine
   c1out << 
    _("Usage: apt [options] command\n"
      "\n"
      "CLI for apt.\n"
      "Basic commands: \n"
      " list - list packages based on package names\n"
      " search - search in package descriptions\n"
      " show - show package details\n"
      "\n"
      " update - update list of available packages\n"
      "\n"
      " install - install packages\n"
      " remove  - remove packages\n"
      "\n"
      " upgrade - upgrade the system by installing/upgrading packages\n"
      " full-upgrade - upgrade the system by removing/installing/upgrading packages\n"
      "\n"
      " edit-sources - edit the source information file\n"
       );
   
   return true;
}

int main(int argc, const char *argv[])					/*{{{*/
{
   CommandLine::Dispatch Cmds[] = {
                                   // query
                                   {"list",&DoList},
                                   {"search", &FullTextSearch},
                                   {"show", &APT::Cmd::ShowPackage},

                                   // package stuff
                                   {"install",&DoInstall},
                                   {"remove", &DoInstall},
                                   {"purge", &DoInstall},

                                   // system wide stuff
                                   {"update",&DoUpdate},
                                   {"upgrade",&DoUpgrade},
                                   {"full-upgrade",&DoDistUpgrade},
                                   // for compat with muscle memory
                                   {"dist-upgrade",&DoDistUpgrade},

                                   // misc
                                   {"edit-sources",&EditSources},

                                   // helper
                                   {"moo",&DoMoo},
                                   {"help",&ShowHelp},
                                   {0,0}};

   std::vector<CommandLine::Args> Args = getCommandArgs("apt", CommandLine::GetCommand(Cmds, argc, argv));

   InitOutput();

   // Set up gettext support
   setlocale(LC_ALL,"");
   textdomain(PACKAGE);

    if(pkgInitConfig(*_config) == false) 
    {
        _error->DumpErrors();
        return 100;
    }

    // some different defaults
   _config->CndSet("DPkg::Progress-Fancy", "1");
   _config->CndSet("Apt::Color", "1");
   _config->CndSet("APT::Get::Upgrade-Allow-New", true);

   // Parse the command line and initialize the package library
   CommandLine CmdL(Args.data(), _config);
   if (CmdL.Parse(argc, argv) == false ||
       pkgInitSystem(*_config, _system) == false)
   {
      _error->DumpErrors();
      return 100;
   }

   if(!isatty(STDOUT_FILENO) && 
      _config->FindB("Apt::Cmd::Disable-Script-Warning", false) == false)
   {
      std::cerr << std::endl
                << "WARNING: " << argv[0] << " "
                << "does not have a stable CLI interface yet. "
                << "Use with caution in scripts."
                << std::endl
                << std::endl;
   }

   // See if the help should be shown
   if (_config->FindB("help") == true ||
       _config->FindB("version") == true ||
       CmdL.FileSize() == 0)
   {
      ShowHelp(CmdL);
      return 0;
   }

   // see if we are in simulate mode
   CheckSimulateMode(CmdL);

   // parse args
   CmdL.DispatchArg(Cmds);

   // Print any errors or warnings found during parsing
   bool const Errors = _error->PendingError();
   if (_config->FindI("quiet",0) > 0)
      _error->DumpErrors();
   else
      _error->DumpErrors(GlobalError::DEBUG);
   return Errors == true ? 100 : 0;
}
									/*}}}*/
